xdem/spatialstats.py

```diff
@@ -1992,28 +2057,242 @@ def _distance_latlon(tup1: tuple, tup2: tuple, earth_rad: float = 6373000) -> fl
+    :param verbose: Print statement to console
+    :param return_in_patch_statistics: Whether to return the dataframe of statistics for all patches and areas
+
+
+    :return: Statistic between patches, Number of patches, Exact discretized area, (Optional) Dataframe of per-patch
+        statistics
+    """
+
+    # Get kernel size to match area
+    # If circular, it corresponds to the diameter
+    if patch_shape.lower() == 'circular':
+        kernel_size = int(np.round(2 * np.sqrt(area / np.pi) / gsd, decimals=0))
+    # If square, to the side length
+    elif patch_shape.lower() == 'square':
+        kernel_size = int(np.round(np.sqrt(area) / gsd, decimals=0))
+
+    else:
+        raise ValueError('Kernel shape should be "square" or "circular".')
+
+    if verbose:
+        print('Computing the convolution on the entire array...')
+    mean_img, nb_valid_img, nb_pixel_per_kernel = mean_filter_nan(img=values, kernel_size=kernel_size,
+                                                                  kernel_shape=patch_shape, method=method)
+
+    # Exclude mean values if number of valid pixels is less than a percentage of the kernel size
+    mean_img[nb_valid_img < nb_pixel_per_kernel * perc_min_valid / 100.] = np.nan
+
+    # A problem with the convolution method compared to the cadrant one is that patches are not independent, which
+    # can bias the estimation of spread. To remedy this, we compute spread statistics on patches separated by the
+    # kernel size (i.e., the diameter of the circular patch, or side of the square patch) to ensure no dependency
+
+    # There are as many combinations for this calculation as the square of the kernel_size
+    if verbose:
+        print('Computing statistic between patches for all independent combinations...')
+    list_statistic_estimates = []
+    list_nb_independent_patches = []
+    for i in range(kernel_size):
+        for j in range(kernel_size):
+            statistic = statistic_between_patches(mean_img[i::kernel_size, j::kernel_size].ravel())
+            nb_patches = np.count_nonzero(np.isfinite(mean_img[i::kernel_size, j::kernel_size]))
+            list_statistic_estimates.append(statistic)
+            list_nb_independent_patches.append(nb_patches)
+
+    if return_in_patch_statistics:
+        # Create dataframe of independent patches for one independent setting
+        df = pd.DataFrame(data={'nanmean' : mean_img[::kernel_size, ::kernel_size].ravel(),
+                                'count': nb_valid_img[::kernel_size, ::kernel_size].ravel()})
+
+    # We then use the average of the statistic computed for different sets of independent patches to get a more robust
+    # estimate
+    average_statistic = np.nanmean(np.asarray(list_statistic_estimates))
+    nb_independent_patches = np.nanmean(np.asarray(list_nb_independent_patches))
+    exact_area = nb_pixel_per_kernel * gsd**2
+
+    if return_in_patch_statistics:
```

Comment:
Some fun overloading when/if we enforce mypy in the future 😉

Reply:
Yeah... :sweat: